### ISSUE TITLE
Wakelock API iOS 16.4 browser mode vs Add to home screen mode

### ISSUE THREAD

**Issue by pelord** [richtr/NoSleep.js#156]
* Device Name: Apple
* Device Version: iOS 16.4
* Browser Name: Safari
* Browser Version: 16.4
* NoSleep.js Version: 0.12.0

### Expected Behavior
NoSleep is suposed to handle iOS 16.4

### Actual Behavior
When using the Wake Lock API, on Safari, in a browser mode, the Wake Lock API work properly.
On the same website, opened after have done "Add to home screen", the Wake Lock API don't work properly. The Wake Lock API seem to not be able to be acquired.

An issue has been reported to WebKit https://bugs.webkit.org/show_bug.cgi?id=205104

### How Do We Reproduce?

Here the test page: https://pelord.github.io/wakeLockApiAddHomeScreen/

To test the Wake Lock API, you have to toggle the ZZZ button.

Unchecked ZZZ = Wake Lock API released
Checked ZZZ = Wake Lock API acquired

I'll try do propose a PR soon.

**Comment by z-x** [comment]
When I saw Safari upgrading to support Wakelock I was like 'AT LAST'. And then I saw it doesn't work in PWA at all, and on top of that it breaks the current NoSleep workaround so my app just broke for people overnight. That is just stupid. Thanks for letting the Safari team know!

**Comment by pelord** [comment]
Here my local fix.

![image](https://user-images.githubusercontent.com/7397743/234319346-0540356c-cee7-4b49-b142-8b4e4c1561b2.png)

**Comment by z-x** [comment]
@pelord - this will disable the Wake Lock for both - in-browser and apps pinned to home screen, right? You could test against this as well to target just the PWAs:

```
window.matchMedia('(display-mode: standalone)').matches
```

**Comment by pelord** [comment]
You're right. In my specific use case, my users are strongly encouraged to "Install" the app... but, I will try it soon. Thanks for the hint..

**Comment by Mrc527** [comment]
I’m having the same issue, in-browser app works but as soon as I install it on my device, the wake lock is not acquired anymore.

Any chance to incorporate the fix a new revision of the library? I know it’s not nice but if the native functionality is broken in PWA mode, we cannot rely on that.

**Comment by pelord** [comment]
Still not fixed in iOS 17
https://bugs.webkit.org/show_bug.cgi?id=254545

**Comment by Mrc527** [comment]
Anything we can do to overcome the issue at least temporarily?

**Comment by jebabin** [comment]
The fix provided by @pelord works perfectly, so I would suggest to use it. When IOS will have a fix, then we will need to chance the regex to only match the bugged versions

**Comment by Mrc527** [comment]
> The fix provided by @pelord works perfectly, so I would suggest to use it. When IOS will have a fix, then we will need to chance the regex to only match the bugged versions

OK but that’s requiring a change in the library, isn’t it? I mean, I’ve to insert nosleep as code instead of installing it via yarn and apply this change on the nosleep code right?

**Comment by robertvanlienden** [comment]
I think https://github.com/Uriopass/NoSleep.js is the most active/working for for this repo.

Maybe we can submit a PR to that repo to add the fix of pelord.
It seems that the owner of this repository never looks back on it :cry: .

For Android this repository works fine on most browsers, but iOS seems to be a pain in the :cake: ...

**Comment by zakj** [comment]
I forked this repository and included the above fix (with @z-x's modification): https://github.com/zakj/no-sleep

It's published on npm as `@zakj/no-sleep`. I couldn't resist some refactoring so buyer beware, but it's working for me.

**Comment by Mrc527** [comment]
Well done, I’m going to switch to your library immediately.I see you even implemented a check to use the old-fashioned way only in the installed PWA, so on a regular browser it should use the native api right?On 13 Oct 2023, at 07:41, Zak Johnson ***@***.***> wrote:﻿
I forked this repository and included the above fix (with @z-x's modification): https://github.com/zakj/no-sleep
It's published on npm as @zakj/no-sleep. I couldn't resist some refactoring so buyer beware, but it's working for me.

**Comment by HenriqueMatias** [comment]
> I forked this repository and included the above fix (with @z-x's modification): https://github.com/zakj/no-sleep
> 
> It's published on npm as `@zakj/no-sleep`. I couldn't resist some refactoring so buyer beware, but it's working for me.

Hey thanks for the improvement in the lib. So this should fix the issue with browsers in IOs 16 ? or I misunderstood?

**Comment by zakj** [comment]
Yep, it's working for me.

**Comment by pelord** [comment]
Here my new fix.

![image](https://github.com/user-attachments/assets/973ed32d-088b-4561-b452-1791626bcc1c)

If the user is on IOS, on a browser, the native wakelock will work. The video playback is strictly used on standalone mode.

**Comment by zakj** [comment]
iOS >= 18.4 [supports wake lock](https://developer.apple.com/documentation/safari-release-notes/safari-18_4-release-notes#Home-Screen-Web-Apps) in PWA mode! I just published [0.13.6](https://www.npmjs.com/package/@zakj/no-sleep) of [my no-sleep fork](https://github.com/zakj/no-sleep) with updated logic to reflect that (https://github.com/zakj/no-sleep/commit/2d1cf259d123af2e7a3375e802598eabb70a88be).
